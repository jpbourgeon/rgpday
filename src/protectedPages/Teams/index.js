import React from 'react'
import { navigate } from '@reach/router'
// import { navigate, Redirect } from '@reach/router'
import PropTypes from 'prop-types'
import { withStyles } from '@material-ui/core/styles'
import Paper from '@material-ui/core/Paper'
import Breadcrumbs from '@material-ui/lab/Breadcrumbs'
import Avatar from '@material-ui/core/Avatar'
import toMaterialStyle from 'material-color-hash'
import Typography from '@material-ui/core/Typography'
import Grid from '@material-ui/core/Grid'
import Divider from '@material-ui/core/Divider'
import IconButton from '@material-ui/core/IconButton'
import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'
import DialogTitle from '@material-ui/core/DialogTitle'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogActions from '@material-ui/core/DialogActions'
import Dialog from '@material-ui/core/Dialog'
import Fab from '@material-ui/core/Fab'
import Game from '@material-ui/icons/VideogameAssetOutlined'
import Search from '@material-ui/icons/Search'
import Close from '@material-ui/icons/Close'
import Delete from '@material-ui/icons/Delete'
import Play from '@material-ui/icons/PlayCircleFilled'
import Edit from '@material-ui/icons/Edit'
import NavigateNext from '@material-ui/icons/NavigateNext'
import Add from '@material-ui/icons/Add'
import Link from 'src/components/Link'
import MUILink from '@material-ui/core/Link'
import Tile from 'src/components/Tile'
import { deleteTeam } from 'src/graphql/mutations'
import logger from 'src/logger'
import API, { graphqlOperation } from '@aws-amplify/api'
import config from 'src/aws-exports'

API.configure(config)
const listTeams = `query ListTeams(
  $filter: ModelTeamFilterInput
  $limit: Int
  $nextToken: String
) {
  listTeams(filter: $filter, limit: $limit, nextToken: $nextToken) {
    items {
      id
      name
      initials
      searchable
    }
    nextToken
  }
}
`

const styles = theme => {
  return {
    '@global': {
      '.grecaptcha-badge': {
        visibility: 'hidden'
      }
    },
    layout: {
      width: 'auto',
      marginLeft: theme.spacing.unit * 3,
      marginRight: theme.spacing.unit * 3,
      [theme.breakpoints.up(1100 + theme.spacing.unit * 3 * 2)]: {
        width: 1100,
        marginLeft: 'auto',
        marginRight: 'auto'
      }
    },
    paper: {
      marginBottom: theme.spacing.unit * 4,
      padding: theme.spacing.unit * 4
    },
    breadcrumb: {
      marginBottom: theme.spacing.unit * 2
    },
    title: {
      marginBottom: theme.spacing.unit * 2
    },
    titleIcon: {
      fontSize: '2.5rem',
      marginBottom: '-0.5rem'
    },
    fab: {
      [theme.breakpoints.up('sm')]: {
        float: 'right'
      },
      [theme.breakpoints.down('xs')]: {
        position: 'fixed',
        bottom: theme.spacing.unit,
        right: theme.spacing.unit
      }
    },
    grow: {
      flexGrow: 1
    },
    loadMoreLink: {
      cursor: 'pointer'
    },
    loadMoreItem: {
      textAlign: 'center'
    },
    hide: {
      display: 'none'
    },
    formContainer: {
      marginBottom: theme.spacing.unit * 2
    },
    filterButtonsContainer: {
      display: 'inline-flex',
      flexDirection: 'column',
      textAlign: 'center',
      justifyContent: 'center'
    },
    filterButtons: {
      padding: theme.spacing.unit,
      marginTop: theme.spacing.unit
    },
    dialog: {
      padding: '1em'
    },
    avatar: {
      margin: 'auto',
      width: 50,
      height: 50,
      fontSize: theme.typography.h6.fontSize,
      color: theme.palette.grey['700'],
      backgroundColor: theme.palette.background.default
    },
    tile: {
      textAlign: 'center',
      justifyContent: 'flex-start'
    },
    tileTitle: {
      marginTop: '1em !important',
      marginBottom: '-0.5em !important'
    },
    tileAvatar: {
      height: '100%'
    },
    tileActions: {
      padding: '0.5em !important',
      marginBottom: '-1em !important'
    },
    itemActionIcon: {
      width: 18,
      height: 18
    },
    itemLargeActionIcon: {
      width: 36,
      height: 36
    },
    itemActionDivider: {
      marginBottom: '0.5em'
    }
  }
}

class Component extends React.Component {
  constructor (props) {
    super(props)
    this._isMounted = false
    this.defaultState = {
      items: [],
      limit: 10,
      nextToken: null,
      filter: '',
      delete: {
        id: '',
        name: '',
        openDialog: false
      }
    }
    this.state = this.defaultState
    this.loadItems = this.loadItems.bind(this)
    this.handleChangeFilter = this.handleChangeFilter.bind(this)
    this.handleCancelFilter = this.handleCancelFilter.bind(this)
    this.handleDeleteDialog = this.handleDeleteDialog.bind(this)
    this.deleteTeam = this.deleteTeam.bind(this)
  }

  componentDidMount () {
    this._isMounted = true
    this.loadItems()
  }

  componentWillUnmount () {
    this._isMounted = false
  }

  async loadItems (event = { preventDefault: () => {} }) {
    event.preventDefault()
    try {
      const state = this.state
      if (state.nextToken === null) state.items = []
      let filter = null
      if (this.state.filter !== '') {
        filter = { searchable: { contains: this.state.filter.toLowerCase() } }
      }
      // GraphQL
      const result = await API.graphql(
        graphqlOperation(listTeams, {
          filter,
          limit: this.state.limit,
          nextToken: this.state.nextToken
        })
      )
      if (!result.errors && result.data.listTeams) {
        state.nextToken = result.data.listTeams.nextToken
        state.items = [...state.items, ...result.data.listTeams.items]
        if (this._isMounted) {
          this.setState(state, () => {
            if (result.data.listTeams.items.length === 0 && result.data.listTeams.nextToken !== null) {
              this.loadItems()
            }
          })
        }
      } else {
        logger.error('loadItems', result)
      }
    } catch (error) {
      logger.error('loadItems', error)
    }
  }

  handleChangeFilter (event) {
    const state = this.state
    state.filter = event.target.value
    this.setState(state)
  }

  handleCancelFilter (event) {
    event.preventDefault()
    this.setState(this.defaultState, () => {
      this.loadItems()
    })
  }

  handleDeleteDialog (id, name, openDialog) {
    this.setState({
      delete: {
        id,
        name,
        openDialog
      }
    })
  }

  async deleteTeam () {
    try {
      // GraphQL
      const id = this.state.delete.id
      if (this._isMounted) this.handleDeleteDialog('', '', false)
      const result = await API.graphql(
        graphqlOperation(deleteTeam, { input: { id } })
      )
      if (!result.errors) {
        const items = this.state.items
        const index = items.findIndex(item => item.id === id)
        if (index > -1) {
          items.splice(index, 1)
        }
        if (this._isMounted) this.setState({ items })
      } else {
        logger.error('deleteTeam', result)
      }
    } catch (error) {
      logger.error('deleteTeam', error)
    }
  }

  render () {
    const { classes, config } = this.props
    // if (!config.sessionId) return (<Redirect noThrow to='/dashboard' />)
    const renderTitle = (item) => (
      <Grid container spacing={40} className={classes.tile}>
        <Grid item xs={12} component='span'>
          <Avatar className={classes.avatar} style={toMaterialStyle(item.name)} >{
            (!item.initials) ? '?' : item.initials
          }</Avatar>
          <div className={classes.tileTitle}>{item.name}</div>
        </Grid>
        <Grid item xs={12} component='span' className={classes.tileActions}>
          <Divider className={classes.itemActionDivider} />
          <div style={{ float: 'left' }}>
            <MUILink onClick={() => { navigate(`/dashboard/serious-game/board/${item.id}`) }}>
              <IconButton><Play color='secondary' className={classes.itemLargeActionIcon} /></IconButton>
            </MUILink>
          </div>
          <div style={{ float: 'right', paddingTop: 9 }}>
            <Link to={`./update-team/${item.id}`}>
              <IconButton><Edit className={classes.itemActionIcon} /></IconButton>
            </Link>
            <MUILink
              onClick={() => { this.handleDeleteDialog(item.id, item.name, true) }}
              className={(config.isAdmin) ? null : classes.hide}
            >
              <IconButton><Delete className={classes.itemActionIcon} /></IconButton>
            </MUILink>
          </div>
        </Grid>
      </Grid>
    )
    const renderItems = (items) => {
      return (
        items.map((item) => {
          return (
            <Tile
              xs={12}
              sm={6}
              md={4}
              key={item.id}
              title={renderTitle(item)}
            />
          )
        })
      )
    }
    return (
      <div className={classes.layout}>
        <main>
          <Paper className={classes.paper}>
            <Grid container className={classes.container}>
              <Grid item xs={12} className={classes.breadcrumb}>
                <Breadcrumbs separator={<NavigateNext fontSize='small' />} arial-label='Breadcrumb'>
                  <Link to='/' color='inherit'>Accueil</Link>
                  <Link to='../' color='inherit'>Tableau de bord</Link>
                  <Link to='.' color='inherit'>Serious Game : choix de l'Ã©quipe</Link>
                </Breadcrumbs>
              </Grid>
              <Grid item xs={12} className={classes.title}>
                <Typography variant='h4' gutterBottom>
                  <Game className={classes.titleIcon} />&nbsp;Serious Game
                  <div className={classes.fab}>
                    <Link to='./add-team'>
                      <Fab color='primary' aria-label='Ajouter'>
                        <Add />
                      </Fab>
                    </Link>
                  </div>
                </Typography>
                <Divider className={classes.title} />
                <Typography variant='h5' gutterBottom>
                  Choisissez votre Ã©quipe ou crÃ©ez-en une nouvelle
                </Typography>
              </Grid>
              <Grid item xs={12}>
                <form
                  className={classes.formContainer}
                  noValidate
                  autoComplete='off'
                  onSubmit={(e) => { this.loadItems(e, true) }}
                >
                  <Grid container>
                    <Grid item className={classes.grow}>
                      <TextField
                        label='Filtrer'
                        fullWidth
                        margin='normal'
                        variant='outlined'
                        onChange={(e) => this.handleChangeFilter(e)}
                        value={this.state.filter}
                      />
                    </Grid>
                    <Grid item className={classes.filterButtonsContainer}>
                      <div className={classes.filterButtons}>
                        <IconButton
                          type='submit'
                        >
                          <Search />
                        </IconButton>
                        <IconButton
                          onClick={this.handleCancelFilter}
                        >
                          <Close />
                        </IconButton>
                      </div>
                    </Grid>
                  </Grid>
                </form>
              </Grid>
              <Grid item xs={12}>
                <Grid container spacing={40} alignItems='baseline'>
                  {(this.state.items.length > 0) ? renderItems(this.state.items) : (
                    <Grid item xs={12}>
                      <Typography variant='body1' gutterBottom>Il n'y a pas d'Ã©quipe Ã  afficher</Typography>
                    </Grid>
                  )}
                  <Grid item xs={12} className={(this.state.nextToken === null) ? classes.hide : null}>
                    <Typography variant='body1' gutterBottom className={classes.loadMoreItem}>
                      <MUILink
                        onClick={this.loadItems}
                        className={classes.loadMoreLink}>
                          Afficher plus de rÃ©sultats
                      </MUILink>
                    </Typography>
                  </Grid>
                </Grid>
              </Grid>
            </Grid>
          </Paper>
        </main>
        <Dialog
          open={this.state.delete.openDialog}
          onClose={() => { this.handleDeleteDialog('', '', false) }}
          aria-labelledby='alert-dialog-title'
          aria-describedby='alert-dialog-description'
        >
          <DialogTitle id='alert-dialog-title'>Effacer l'Ã©quipe {this.state.delete.name}</DialogTitle>
          <DialogContent>
            <DialogContentText id='alert-dialog-description'>
              Attention, cette action est irrÃ©versible.
            </DialogContentText>
          </DialogContent>
          <DialogActions className={classes.dialog}>
            <Button onClick={this.deleteTeam} color='secondary' variant='contained'>
              Confirmer
            </Button>
            <Button onClick={() => { this.handleDeleteDialog('', '', false) }} color='primary' autoFocus variant='outlined'>
              Annuler
            </Button>
          </DialogActions>
        </Dialog>
      </div>
    )
  }
}

Component.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Component)
