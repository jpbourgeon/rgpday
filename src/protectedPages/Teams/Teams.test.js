/* global describe, it, expect */
import React from 'react'
import renderer from 'react-test-renderer'
import Teams from './'
import jssSerializer from 'src/utils/jssSnapshotSerializer'

expect.addSnapshotSerializer(jssSerializer)

describe('The Teams component', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Teams />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
