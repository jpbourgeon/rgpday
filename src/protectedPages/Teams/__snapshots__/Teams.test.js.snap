// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Teams component should render correctly 1`] = `
<div
  className="layout-1"
>
  <main>
    <div
      className="root-22 elevation2-26 rounded-23 paper-2"
    >
      <div
        className="container-49"
      >
        <div
          className="item-50 12-89 breadcrumb-3"
        >
          <nav
            arial-label="Breadcrumb"
            className="root-150 body1-159 colorTextSecondary-183 root-146"
          >
            <ol
              className="ol-147"
            >
              <li
                className="li-148"
              >
                <a
                  aria-current="page"
                  className="root-150 colorInherit-179 inline-185 root-186 underlineHover-188"
                  href="/"
                  onClick={[Function]}
                >
                  Accueil
                </a>
              </li>
              <li
                aria-hidden="true"
                className="root-191 separator-149"
              >
                <svg
                  aria-hidden="true"
                  className="root-192 fontSizeSmall-199"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </li>
              <li
                className="li-148"
              >
                <a
                  aria-current="page"
                  className="root-150 colorInherit-179 inline-185 root-186 underlineHover-188"
                  href="/"
                  onClick={[Function]}
                >
                  Tableau de bord
                </a>
              </li>
              <li
                aria-hidden="true"
                className="root-191 separator-149"
              >
                <svg
                  aria-hidden="true"
                  className="root-192 fontSizeSmall-199"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </li>
              <li
                className="li-148"
              >
                <a
                  aria-current="page"
                  className="root-150 colorInherit-179 inline-185 root-186 underlineHover-188"
                  href="/"
                  onClick={[Function]}
                >
                  Serious Game : choix de l'équipe
                </a>
              </li>
            </ol>
          </nav>
        </div>
        <div
          className="item-50 12-89 title-4"
        >
          <h4
            className="root-150 h4-165 gutterBottom-177"
          >
            <svg
              aria-hidden="true"
              className="root-192 titleIcon-5"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <g>
                <path
                  d="M21 6H3c-1.1 0-2 .9-2 2v8c0 1.1.9 2 2 2h18c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2zm0 10H3V8h18v8z"
                />
                <path
                  d="M6 15h2v-2h2v-2H8V9H6v2H4v2h2z"
                />
                <circle
                  cx="14.5"
                  cy="13.5"
                  r="1.5"
                />
                <circle
                  cx="18.5"
                  cy="10.5"
                  r="1.5"
                />
              </g>
            </svg>
             Serious Game
            <div
              className="fab-6"
            >
              <a
                className="root-150 colorPrimary-180 inline-185 root-186 underlineHover-188"
                href="//add-team"
                onClick={[Function]}
              >
                <button
                  aria-label="Ajouter"
                  className="root-211 root-201 primary-203"
                  disabled={false}
                  onBlur={[Function]}
                  onContextMenu={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex="0"
                  type="button"
                >
                  <span
                    className="label-202"
                  >
                    <svg
                      aria-hidden="true"
                      className="root-192"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                      />
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                      />
                    </svg>
                  </span>
                  <span
                    className="root-295"
                  />
                </button>
              </a>
            </div>
          </h4>
          <hr
            className="root-214 title-4"
          />
          <h5
            className="root-150 h5-166 gutterBottom-177"
          >
            Choisissez votre équipe ou créez-en une nouvelle
          </h5>
        </div>
        <div
          className="item-50 12-89"
        >
          <form
            autoComplete="off"
            className="formContainer-11"
            noValidate={true}
            onSubmit={[Function]}
          >
            <div
              className="container-49"
            >
              <div
                className="item-50 grow-7"
              >
                <div
                  className="root-219 marginNormal-220 fullWidth-222"
                >
                  <label
                    className="root-234 root-223 formControl-228 animated-231 outlined-233"
                    data-shrink={false}
                  >
                    Filtrer
                  </label>
                  <div
                    className="root-254 root-241 fullWidth-263 formControl-255"
                    onClick={[Function]}
                  >
                    <fieldset
                      aria-hidden={true}
                      className="root-271 notchedOutline-248"
                      style={
                        Object {
                          "paddingLeft": 8,
                        }
                      }
                    >
                      <legend
                        className="legend-272"
                        style={
                          Object {
                            "width": 0.01,
                          }
                        }
                      >
                        <span
                          dangerouslySetInnerHTML={
                            Object {
                              "__html": "&#8203;",
                            }
                          }
                        />
                      </legend>
                    </fieldset>
                    <input
                      aria-invalid={false}
                      className="input-264 input-249"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      required={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                className="item-50 filterButtonsContainer-12"
              >
                <div
                  className="filterButtons-13"
                >
                  <button
                    className="root-211 root-273"
                    disabled={false}
                    onBlur={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="0"
                    type="submit"
                  >
                    <span
                      className="label-278"
                    >
                      <svg
                        aria-hidden="true"
                        className="root-192"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                        />
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                      </svg>
                    </span>
                    <span
                      className="root-295"
                    />
                  </button>
                  <button
                    className="root-211 root-273"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex="0"
                    type="button"
                  >
                    <span
                      className="label-278"
                    >
                      <svg
                        aria-hidden="true"
                        className="root-192"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                      </svg>
                    </span>
                    <span
                      className="root-295"
                    />
                  </button>
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          className="item-50 12-89"
        >
          <div
            className="container-49 40-75 baseline-60"
          >
            <div
              className="item-50 12-89"
            >
              <p
                className="root-150 body1-159 gutterBottom-177"
              >
                Il n'y a pas d'équipe à afficher
              </p>
            </div>
            <div
              className="item-50 12-89 hide-10"
            >
              <p
                className="root-150 body1-159 gutterBottom-177 loadMoreItem-9"
              >
                <a
                  className="root-150 colorPrimary-180 inline-185 root-186 underlineHover-188 loadMoreLink-8"
                  onClick={[Function]}
                >
                  Afficher plus de résultats
                </a>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
